### PR TITLE
Pure go reimplementation

### PR DESCRIPTION
**Pull request by spheromak**
Moving away from libgit to somehting more portable.  Also fixing a bug where major doesn't set the minor and patch to 0 when it bumps

---
This change is [Reviewable](https://reviewable.io/reviews/pantheon-systems/autotag/4)

### CODE REVIEW

File: autotag.go [hunk 1]
```diff
@@ -132,28 +137,58 @@ func (r *GitRepo) LatestVersion() string {
...
-		if err != nil {
+	l, err := r.Repo.CommitsBefore(r.branchID)
+	if err != nil {
+		log.Printf("Error loading history for tag '%s': %s ", r.currentVersion, err.Error())
+	}
+	log.Printf("Checking commits from %s to %s ", r.branchID, tagCommit.Id)
+
+	start := false
+	for e := l.Back(); e != nil; e = e.Prev() {
+		commit := e.Value.(*git.Commit)
```
Comment (kf6nux) [inline autotag.go]: Can this ever be either another type or return a nil pointer?

File: autotag.go [hunk 2]
```diff
@@ -132,28 +137,58 @@ func (r *GitRepo) LatestVersion() string {
...
-		v, err := r.parseCommit(c)
-		if err != nil {
+	l, err := r.Repo.CommitsBefore(r.branchID)
+	if err != nil {
+		log.Printf("Error loading history for tag '%s': %s ", r.currentVersion, err.Error())
+	}
+	log.Printf("Checking commits from %s to %s ", r.branchID, tagCommit.Id)
+
+	start := false
+	for e := l.Back(); e != nil; e = e.Prev() {
```
Comment (kf6nux) [inline autotag.go]: This might be comment worthy.  Something like,
> Chronologically (oldest to newest) process commit history

File: autotag.go [hunk 3]
```diff
@@ -165,36 +200,19 @@ func (r *GitRepo) calcVersion() error {
...
-	if err := r.tagNewVersion(); err != nil {
-		return err
-	}
-
-	return nil
+	return r.tagNewVersion()
 }
 
 func (r *GitRepo) tagNewVersion() error {
 	// TODO:(jnelson) These should be configurable? Mon Sep 14 12:02:52 2015
```
Comment (kf6nux) [inline autotag.go]: Does this comment still apply?

File: autotag.go [hunk 4]
```diff
@@ -203,7 +221,7 @@ func (r *GitRepo) tagNewVersion() error {
 func (r *GitRepo) parseCommit(commit *git.Commit) (*version.Version, error) {
 	var b bumper
 	msg := commit.Message()
-	log.Println("Parsing ", msg, ":")
+	log.Printf("Parsing %s: %s", commit.Id, msg)
```
Comment (kf6nux) [inline autotag.go]: Do we want a `\n` in the format string?  (this comment might apply to multiple output messages in this repo).

File: autotag.go [hunk 5]
```diff
@@ -58,56 +55,64 @@ func NewRepo(repoPath string) (*GitRepo, error) {
...
 
-	cRange := fmt.Sprintf("%s..HEAD", lastParsedTagRef)
-	//	log.Println("Check range:", cRange)
-	if wErr := r.walker.PushRange(cRange); wErr != nil {
-		log.Fatal(wErr)
+	keys := make([]*version.Version, 0, len(versions))
+	for key := range versions {
+		keys = append(keys, key)
 	}
+	sort.Sort(version.Collection(keys))
```
Comment (spheromak) [inline autotag.go]: reverse returns a new slice, I was just reusing it.

File: autotag.go [hunk 6]
```diff
@@ -58,56 +55,64 @@ func NewRepo(repoPath string) (*GitRepo, error) {
...
 
-	sort.Sort(version.Collection(versions))
+	// set the current versions
+	if itemLen := len(keys); itemLen >= 1 {
```
Comment (spheromak) [inline autotag.go]: then we would need to get the len again to do the end of slice element we get in the following line. Can be simplified with the reverse sort. need to look at it all in that context tho.

File: autotag.go [hunk 7]
```diff
@@ -58,56 +55,64 @@ func NewRepo(repoPath string) (*GitRepo, error) {
...
 		}
 	}
 
-	cRange := fmt.Sprintf("%s..HEAD", lastParsedTagRef)
-	//	log.Println("Check range:", cRange)
-	if wErr := r.walker.PushRange(cRange); wErr != nil {
-		log.Fatal(wErr)
+	keys := make([]*version.Version, 0, len(versions))
+	for key := range versions {
+		keys = append(keys, key)
```
Comment (spheromak) [inline autotag.go]: how else would you pull a slice of the keys here ? we want to sort them.

**Comment by spheromak** [comment]
@kf6nux addressed most the comments except the append stuff  since I am not sure what you are saying we should do there.

File: autotag.go (hunk 7 above)
Comment (kf6nux) [inline autotag.go]:
```keys := make([]*version.Version, len(versions))
i := 0
for key := range versions {
	keys[i] =  key)
	i++```

**Pull request merged**